[feature/client-create-delete-follow] Client 팔로우 Create/Delete API 구현

HanJuK opened the pull request:
## 개요
### PR 종류
**해당되는 한 가지만 선택하세요.**
 - [X] 기능 구현을 위한 소스코드 수정 (PR제목 양식: "[feature/브랜치명] 제목", 브랜치작명법: "feature/명칭" 사용)
 - [ ] 버그 수정을 위한 소스코드 수정 (PR제목 양식: "[hotfix/브랜치명] 제목", 브랜치작명법: "hotfix/명칭" 사용)
 - [ ] 소스코드 수정을 동반하지 않는 단순 문서 편집 (PR제목 양식: "[docs/브랜치명] 제목", 브랜치작명법: "docs/명칭" 사용)

### 작업내용
안드로이드 Client의 팔로우 Create/Delete API를 구현하였습니다.

## 변경사항
- 커뮤니티 프래그먼트의 ActionBar에 팔로워/팔로잉 버튼 추가
- 팔로워/팔로잉 액티비티 및 프래그먼트 생성
    - 유저 검색 버튼 생성(이번 PR 주 작업 내용)
    - ViewPager2를 이용하여 팔로워, 팔로잉 프래그먼트를 각각 담음(추후 작업 예정)
- 유저 검색 액티비티 생성
    - 프래그먼트 대신 Dialog 테마의 액티비티로 구현(구조가 매우 간단하고, 독립적인 레이아웃이기 때문)
    - 닉네임(nickname)으로 사용자를 검색하는 방식으로 구현(서버에 관련 로직 추가)
    - 본인(self)의 닉네임을 검색하는 예외 처리 관련 코드 추가(서버)
    - 사용자 id(username 아님)로 프로필 사진 fetch 구현(서버에 관련 로직 추가)
        - **(이 변경사항 때문에 기존에 있던 계정 프로필 사진 fetch 관련 코드가 작동하지 않을 수 있습니다. 이는 제가 추후에 hotfix로 수정하겠습니다.)**
    - 클라이언트 FetchFollower, CreateFollow, DeleteFollow API 로직 구현
    - 액티비티 생명구조 대응
- 기타
    - retrofit onFailure() 내 토스트 메시지 관련 코드 순서 일부 변경(필요한 코드에 도달하지 못하는 경우 있었음)

## 비고
### TODO
- 팔로워/팔로잉 프래그먼트 레이아웃 및 로직 구현

### 스크린샷
- ActionBar에 팔로워/팔로잉 버튼 / 팔로워/팔로잉 레이아웃 / 유저 검색 액티비티 레이아웃
<img src="https://user-images.githubusercontent.com/48195650/129481601-007b7abb-44c3-4400-a1f7-54520d93f695.png" width="30%"> <img src="https://user-images.githubusercontent.com/48195650/129481604-c253ea98-48ab-4626-932a-7b1f6402cb4e.png" width="30%"> <img src="https://user-images.githubusercontent.com/48195650/129481608-94c3a665-9f4a-40ec-89f8-e55bdadc4d36.png" width="30%">

- 팔로우 하지 않은 경우 / 팔로우 한 경우 / 팔로일 버튼 클릭 시 API 결과
<img src="https://user-images.githubusercontent.com/48195650/129481700-0d32d253-d7dd-43ce-8d4a-8b211742f34a.png" width="30%"> <img src="https://user-images.githubusercontent.com/48195650/129481701-14ae8d7a-2829-4f5e-9e8b-3754b8a2a487.png" width="30%"> <img src="https://user-images.githubusercontent.com/48195650/129481703-d0faa2e9-9877-4971-a50b-227d53bfe868.PNG" width="30%">

- 존재하지 않는 닉네임 예외 / 자기 자신 검색 예외
<img src="https://user-images.githubusercontent.com/48195650/129481728-a92d41a7-263d-455c-a967-c55541c6012e.png" width="30%"> <img src="https://user-images.githubusercontent.com/48195650/129481730-844c569c-fec1-4a15-9d64-a0e68d483192.png" width="30%">

## 품질 관리를 위한 체크리스트
 - [ ] 해당사항 없음 (**소스코드 수정을 하지 않았습니다.**)
### 소스코드 테스트
**세 가지 테스트 모두를 실시하는 것을 권장하며, 최소한 작동확인(수동 테스트)은 해보셔야 합니다.**
 - [X] 수동 테스트 / Manual Test (앱 구동후 사용확인을 통한 직접 테스트 또는 서버 구동후 Postman을 이용한 테스트)
 - [ ] 자동화 단위 테스트 / Unit Test (JUnit 등을 통해 개별 단위로직에 대한 테스트 스크립트 작성)
 - [ ] 자동화 통합 테스트 / Integration Test (Spring MVC Testing 등을 통해 프로그램 전체에 대한 테스트 스크립트 작성)
### 기본 확인사항
 - [X] IDE에서 검출되는 모든 경고와 오류를 해결하셨습니까?
 - [X] 코딩 컨벤션(-추후 문서화 예정- 작명법, 금기사항, 구현방식 등)을 준수하셨습니까?
 - [X] 버그 또는 구현상의 실수, 누락 등 하자사항이 없는지 테스트하여 확인해보셨습니까?
 - [X] 명칭(변수명, 함수명, 주석 등)에 오타가 있는지 테스트하여 확인해보셨습니까?
 - [X] 이번 변경사항에 포함된 주석 및 기능 명세, 관련 문서를 최신화하셨습니까?

HanJuK committed:
created necessary files for follower

HanJuK committed:
created layouts for follower fragment

HanJuK committed:
created action bar menu item for follower/following

HanJuK committed:
implemented ViewPager2 for follower following fragment

HanJuK committed:
created files for follower RecyclerView

HanJuK committed:
created layout for search activity

HanJuK committed:
implemented regex inside search activity's EditText

HanJuK committed:
implemented logic for searching for account by nickname

HanJuK committed:
retrofit onFailure Toast message bug related changes

HanJuK committed:
handled exceptions for search(added fetch self check related code inside server)

HanJuK committed:
implemented logic for fetching account photo(server files changed as well)

HanJuK committed:
implemented logic for fetching follower id list

HanJuK committed:
implemented logic for create/delete follow API

Komputer-P commented on server/src/main/java/com/sju18/petmanagement/domain/account/api/AccountController.java:
```diff
@@ -66,12 +77,19 @@
         byte[] fileBinData;
         try {
-            fileBinData = accountServ.fetchAccountPhoto(auth);
+            // if id is null -> fetch self photo
```
Controller는 엔드포인트가 받은 요청을 서비스의 함수만 호출해서 그 결과를 처리하여 다시 보내는 영역이기 때문에 해당 로직은 AccountService에 두는 것을 추천드립니다.

Komputer-P commented on server/src/main/java/com/sju18/petmanagement/domain/account/dao/AccountRepository.java:
```diff
@@ -8,6 +8,7 @@
 @Repository
 public interface AccountRepository extends CrudRepository<Account, Long> {
     Optional<Account> findByUsername(String username);
+    Optional<Account> findByNickname(String username);
```
findByNickname() 함수의 매개변수의 이름을 username -> nickname 로 바꾸는 것은 어떨까요?

Komputer-P commented on server/src/main/java/com/sju18/petmanagement/domain/account/api/AccountController.java:
```diff
@@ -53,6 +53,17 @@
+                // 해당 nickname 가진 계정 정보 조회
+                account = accountServ.fetchAccountByNickname(reqDto.getNickname());
+
+                // check if self -> return exception
```
Controller는 엔드포인트가 받은 요청을 서비스의 함수만 호출해서 그 결과를 처리하여 다시 보내는 영역이기 때문에 해당 로직은 AccountService에 두는 것을 추천드립니다.
또한 logger.wan의 내용은 static/messages/account/error_en.properties에 따로 저장한 후 가져와 사용하는 것을 추천드립니다. (안드로이드 string.xml처럼)

krapie commented on server/src/main/java/com/sju18/petmanagement/domain/account/api/AccountController.java:
```diff
@@ -53,6 +53,17 @@
+                // 해당 nickname 가진 계정 정보 조회
+                account = accountServ.fetchAccountByNickname(reqDto.getNickname());
+
+                // check if self -> return exception
```
Controller는 엔드포인트가 받은 요청을 서비스의 함수만 호출해서 그 결과를 처리하여 다시 보내는 영역이기 때문에 해당 로직은 AccountService에 두는 것을 추천드립니다.
또한 logger.wan의 내용은 static/messages/account/error_en.properties에 따로 저장한 후 가져와 사용하는 것을 추천드립니다. (안드로이드 string.xml처럼)

HanJuK replied:
제가 Controller와 Service의 개념에 대한 이해가 부족했네요.. 관련하여 수정 및 커밋(bef22d94a99560a7449c1b2d79a0d045c6e48a99)했습니다. 말씀하신 error_en.properties 또한 적용했습니다.

krapie commented on server/src/main/java/com/sju18/petmanagement/domain/account/api/AccountController.java:
```diff
@@ -66,12 +77,19 @@
         byte[] fileBinData;
         try {
-            fileBinData = accountServ.fetchAccountPhoto(auth);
+            // if id is null -> fetch self photo
```
Controller는 엔드포인트가 받은 요청을 서비스의 함수만 호출해서 그 결과를 처리하여 다시 보내는 영역이기 때문에 해당 로직은 AccountService에 두는 것을 추천드립니다.

HanJuK replied:
위와 마찬가지로 수정 및 커밋(bcb7e1aa3bd42dd5ecdcf2051a3ac14f7ccb2f61)했습니다.

krapie commented on server/src/main/java/com/sju18/petmanagement/domain/account/dao/AccountRepository.java:
```diff
@@ -8,6 +8,7 @@
 @Repository
 public interface AccountRepository extends CrudRepository<Account, Long> {
     Optional<Account> findByUsername(String username);
+    Optional<Account> findByNickname(String username);
```
findByNickname() 함수의 매개변수의 이름을 username -> nickname 로 바꾸는 것은 어떨까요?

HanJuK replied:
오타가 있었네요.. 수정 및 커밋(efcdbcf2363fcc8a4fee0b44deffaa72379efa31)했습니다. 제보 감사합니다.

krapie reviewed (changes requested):
일단 서버 측에만 코멘트를 달았습니다.
클라이언트는 추후에 하거나, @Hanbit-Kang 님이 한 후에 보도록 하겠습니다.

HanJuK committed:
refactored fetched self related code to AccountService + created error message property

hanbikan commented on client/android/app/src/main/java/com/sju18001/petmanagement/ui/community/followerFollowing/SearchActivity.kt:
```diff
@@ -0,0 +1,475 @@
+                    searchViewModel.apiIsLoading = false
+                    setSearchButtonToNormal()
+
+                    // get error message + show(Toast)
```
자잘한 요청 사항입니다. 에러 메시지를 출력하는 구문이 아래와 같이 '코틀린스러워졌으면' 합니다..

![image](https://user-images.githubusercontent.com/58168528/129507742-1f381f8a-dedd-4604-8a9e-e64b9443a050.png)

완전히 같아야할 필요는 없습니다만 코틀린에서는 위와 같은 코드를 권장하는 듯합니다!

HanJuK replied:
코틀린의 when을 사용한 코드가 훨씬 간결하고 좋네요. 올려주신 코드와 완전 동일하지는 않지만(최대한 기존 코드 형태를 유지했음) 말씀하신 방향대로 수정 및 커밋(f2ee80002cf8858760e702ce04698e4cb68edd73)했습니다.

HanJuK committed:
refactored logic for fetch account photo by id(read description)

- AccountService.java: deleted and refactored fetchAccountPhotoById() into fetchAccountPhoto()
- refactored id null check logic from AccountController.java to AccountService.java

HanJuK committed:
typo fix

HanJuK committed:
minor code cleanup(else if -> when)

HanJuK committed:
changed id's null value from -1 to null(+ read description) + created FetchAccountPhotoReqDto
